app/grandchallenge/algorithms/models.py

```diff
@@ -581,19 +581,46 @@ def run_job(self, upload_pks=None):
         on_commit(run_job.apply_async)
 
     @cached_property
-    def special_outputs(self):
-        outputs = {
-            "CHART": [],
-            "PDF": [],
-        }
+    def special_outputs(self):  # noqa: C901
+        outputs = self.output_interface_dictionary
+
         for output in self.outputs.all():
             if (
                 output.interface.kind
                 == InterfaceKind.InterfaceKindChoices.CHART
             ):
-                outputs["CHART"].append(output)
+                for title in outputs["CHART"]:
+                    if title == output.interface.title:
+                        outputs["CHART"][f"{title}"] = output
             if output.interface.kind == InterfaceKind.InterfaceKindChoices.PDF:
-                outputs["PDF"].append(output)
+                for title in outputs["PDF"]:
+                    if title == output.interface.title:
+                        outputs["PDF"][f"{title}"] = output
+            if output.interface.kind in (
+                InterfaceKind.InterfaceKindChoices.THUMBNAIL_JPG,
+                InterfaceKind.InterfaceKindChoices.THUMBNAIL_PNG,
+            ):
+                for title in outputs["TIMG"]:
+                    if title == output.interface.title:
+                        outputs["TIMG"][f"{title}"] = output
+
+        return outputs
+
+    @cached_property
+    def output_interface_dictionary(self):
+        outputs = {
+            "CHART": {},
+            "PDF": {},
+            "TIMG": {},
+        }
+        for kind, title in self.output_interfaces.values_list("kind", "title"):
```

Comment:
This results in duplicate queries (one for each job), but I cannot think of a way around this right now. First creating a dictionary with all output interfaces that the algorithm has (and only then filling them with the interface values) is necessary (I think) because not all jobs for an algorithm might have values for all configured interfaces, and if we just loop over all existing interface values per job, a given value might end up in the wrong column...

Reply:
Yes, we need to avoid that. Why not fill out the data for each job and then filter when you fill out the columns in the view instead? Also, do not use titles for indexes, always use slugs.